numpy/lib/tests/test_loadtxt.py

```diff
@@ -234,37 +235,88 @@ def test_converters_negative_indices_with_usecols():
     assert_array_equal(res, [[0, -1], [0, -1]])
 
-def test_ragged_usecols():
+
+@pytest.mark.parametrize('usecols', [[0, -2], lambda n: [0, -2]])
+def test_ragged_usecols(usecols):
     # usecols, and negative ones, work even with varying number of columns.
     txt = StringIO("0,0,XXX\n0,XXX,0,XXX\n0,XXX,XXX,0,XXX\n")
     expected = np.array([[0, 0], [0, 0], [0, 0]])
-    res = np.loadtxt(txt, dtype=float, delimiter=",", usecols=[0, -2])
+    res = np.loadtxt(txt, dtype=float, delimiter=",", usecols=usecols)
     assert_equal(res, expected)
 
     txt = StringIO("0,0,XXX\n0\n0,XXX,XXX,0,XXX\n")
     with pytest.raises(ValueError,
                 match="invalid column index -2 at row 2 with 1 columns"):
         # There is no -2 column in the second row:
-        np.loadtxt(txt, dtype=float, delimiter=",", usecols=[0, -2])
+        np.loadtxt(txt, dtype=float, delimiter=",", usecols=usecols)
 
 
-def test_empty_usecols():
+@pytest.mark.parametrize('usecols', [[], lambda n: []])
+def test_empty_usecols(usecols):
     txt = StringIO("0,0,XXX\n0,XXX,0,XXX\n0,XXX,XXX,0,XXX\n")
-    res = np.loadtxt(txt, dtype=np.dtype([]), delimiter=",", usecols=[])
+    res = np.loadtxt(txt, dtype=np.dtype([]), delimiter=",", usecols=usecols)
     assert res.shape == (3,)
     assert res.dtype == np.dtype([])
 
 
+def test_callable_usecols_basic():
+    txt = StringIO('1 2 3 4 5\n6 7 8 9 1\n2 3 4 5 6\n')
+    res = np.loadtxt(txt, dtype=int, usecols=lambda n: list(range(n)[::2]))
+    expected = np.array([[1, 3, 5], [6, 8, 1], [2, 4, 6]])
+    assert_equal(res, expected)
+
+
+def test_callable_usecols_method():
+
+    class Foo:
+
+        def usecols(self, n):
+            return list(range(n)[1::2])
+
+        def load(self):
+            return np.loadtxt(StringIO('1 2 3 4\n5 6 7 8'),
+                              usecols=self.usecols)
+
+    res = Foo().load()
+    expected = np.array([[2.0, 4.0], [6.0, 8.0]])
+    assert_equal(res, expected)
+
+
+def test_callable_usecols_raises_exception():
+    with pytest.raises(ZeroDivisionError, match='division by zero'):
+        np.loadtxt(StringIO('1 2\n4 5'), usecols=lambda n: 1/0)
+
+
+def test_callable_usecols_not_a_sequence():
+    with pytest.raises(TypeError,
+                       match="returned an instance of type 'module'"):
+        np.loadtxt(StringIO('1 2\n4 5'), usecols=lambda n: np)
+
+
+def test_callable_usecols_bad_element_type():
+    with pytest.raises(TypeError,
+                       match="at least one occurrence of type 'module'"):
+        np.loadtxt(StringIO('1 2\n3 4'), usecols=lambda n: [np])
+
+
+def test_callable_usecols_bad_len_with_structured_dtype():
+    dt = np.dtype([('x', float), ('y', float)])
+    txt = StringIO('1 2 3 4\n5 6 7 8\n')
+    with pytest.raises(RuntimeError,
+                       match='the number of fields in the given dtype'):
+        np.loadtxt(txt, usecols=lambda n: [0, 2, 3], dtype=dt)
+
```

Comment:
We are missing a test (and I think also the logic) to reject a change in the number of columns when a callable is given.  This is a bit of a tricky case!

The problem is that if `n` changes, the return value of the `usecol` function might also change!  While normally, when usecols are given we do not have to worry about that at all.

The example would be something like:
```
np.loadtxt(["1 2", "1 2 3"], usecols=lambda n: range(n)[-2:])
```

Reply:
As you can see in the code, I've been working under the assumption that once the first row has been read, the value of `n` is determined once and for all.  If given, the user-defined callable is called just *once*, after the first row is read.   If a structured dtype is given, `n` is the number of fields in the dtype; otherwise `n` is the number of columns found in the first row of the file.  That's because I didn't think we had such strong support for ragged text files.  I wish we didn't, and I don't recall the history of how we got here, but if that's where we are now, so be it.  (There is no formal specification for the format of the text files that we support, so the scope of `loadtxt` tends to creep in fits and starts.)

It seems that a simple way to handle this is to cache the results of the callable `usecols` in a dictionary, and look up the result after each line is parsed.  We only actually call the function when a new number of columns is encountered.  There will have to be a check added to ensure that the length of the sequence returned by subsequent calls is the same as that of the first call.  What do you think?

Reply:
I don't mind raising an error.  We already do if `usecols` is *not* given.  We do have oddly strong support for changing number of columns, simply because `loadtxt` used to use list indexing `for col in usecol: line.split(delimiter)[col]`.

Yes, you could cache `usecols(n)` with changing `n` (in the unlikely event that it chagned).  Or... you just raise an error :).

I would be OK even deprecating that whole thing.  Just dropping support for *negative* usecols together with *ragged* columns, might simplify things also.